Add FILE line to Cuefile to include file path for each track

presentformyfriends opened the pull request:
Fixes [#13321](https://github.com/mixxxdj/mixxx/issues/13321)

This PR adds a FILE line to the Cuefile to show a filepath for each track (see issue for more information).

presentformyfriends committed:
Add FILE line to Cuefile to include file path for each track

acolombier reviewed (commented):
Thanks for the contribution, as @fwcd suggested, it would be great to make that an option. To do so, you could add a boolean to `EngineRecord` (like `m_bCueUsesFileAnnotation`), and initialise it in the `EngineRecord::updateFromPreferences`, using `m_pConfig->getValue()`. 
You can use Qt Designer to edit `src/preferences/dialog/dlgprefrecorddlg.ui` and add the logic in `src/preferences/dialog/dlgprefrecord.cpp` to ahndle the preference edition. Happy to do that part if you would like me to.

ronso0 commented:
Just a word on Qt Designer: that potentially rearranges the .ui files (for no reason) which produces a unreviewable diff.
If that happens please resort to editing the ui file manually in a text editor.

presentformyfriends commented:
@ronso0 Thanks for the tip, I will pay attention to the .ui files order and will edit manually if need be

presentformyfriends commented:
@acolombier Quick question, I just read in the Developer Guidelines to prefer merging over rebasing as rebasing apparently causes problems. I did not know this, and can't remember if I rebased or not. Is there a way for me to check if I rebased? Sorry if this is a dumb question. Thanks!

JoergAtGithub commented:
Welcome at Mixxx!
As a first-time contributor we need you to sign the [Mixxx Contributor Agreement](https://docs.google.com/forms/d/e/1FAIpQLScC9QG327sjLL0eWftmfGUasxFFLxg6LCXJ2xHDYRzFIRqyiw/viewform?formkey=dEpYN2NkVEFnWWQzbkFfM0ZYYUZ5X2c6MQ) and comment here when you have done so. It gives us permission to distribute your contribution under the GPL v2 or later license and the Apple Mac App Store. It is also helpful for us to have contact information for contributors in case we may need it in the future.

ronso0 commented:
@presentformyfriends Merging is preferred becasue rebasing rewrites the commit history and may detach review comments (which could mean a lot / everything hd to be reviewwd again)

Though, for this minimal PR that's irrelevant IMO.

presentformyfriends commented:
Thanks @JoergAtGithub just signed it!

> Welcome at Mixxx! As a first-time contributor we need you to sign the

presentformyfriends commented:
Got it, good to know!

> Merging is preferred becasue rebasing rewrites the commit history and may detach review comments (which could mean a lot / everything hd to be reviewwd again)
> 
> Though, for this minimal PR that's irrelevant IMO.

presentformyfriends committed:
Merge branch 'mixxxdj:main' into fixing_some_bug

presentformyfriends committed:
Add CUE file annotation feature logic and UI

Add boolean to EngineRecord and initialize it in updateFromPreferences

Add checkbox to DlgPrefRecord to enable file annotation in CUE file

Add logic to ensure checkbox is disabled by default, include tooltip

presentformyfriends commented:
Pushed the rest, please let me know if it's ok or if I should change anything. FYI: I edited the .ui files with kDevelop, not Qt Designer.

presentformyfriends committed:
Fix formatting issues as per pre-commit hook

presentformyfriends commented:
Can anyone tell me why my commit keeps failing on clang-format? I ran pre-commit locally and everything passes. When I look at the diff it's formatting issues that I've already fixed? 🤷

Swiftb0y commented:
> when I look at the diff it's formatting issues that I've already fixed? 🤷

Are you sure? The lines that clang-format on CI complaints about missing linebreaks in `enginerecord.cpp:128` for example.

Also keep in mind that pre-commit only looks at the most recent commit, so if you have not run pre-commit since starting the PR, you might need to do a `pre-commit run --from-ref upstream/main --to-ref HEAD` so it runs on the entire PR.

acolombier commented on src/preferences/dialog/dlgprefrecord.cpp:
```diff
@@ -76,6 +76,9 @@ DlgPrefRecord::DlgPrefRecord(QWidget* parent, UserSettingsPointer pConfig)
             ConfigKey(RECORDING_PREF_KEY, "CueEnabled"), kDefaultCueEnabled));
 
+    CheckBoxUseCueFileAnnotation->setChecked(m_pConfig->getValue<bool>(
+            ConfigKey(RECORDING_PREF_KEY, "CueFileAnnotationEnabled"), false));
```
We are trying to streamline new CO and config keys to use `snake_case`
```suggestion
            ConfigKey(RECORDING_PREF_KEY, "cue_file_annotation_enabled"), false));
```

acolombier commented on src/engine/sidechain/enginerecord.cpp:
```diff
@@ -249,6 +251,12 @@ void EngineRecord::writeCueLine() {
                             .toUtf8());
 
+    if (m_bCueUsesFileAnnotation) {
+        m_cueFile.write(QString("    FILE \"%1\"\n")
```
I believe you might be able to use a `QStringLiteral`

presentformyfriends replied:
For this one, fyi all the `m_cueFile.write` lines are existing code and use QString, I just added the if statement on line 254. I assumed whoever wrote that code used QString for a reason but idk.

acolombier replied:
`QStringLiteral` is a `QString` :) ([doc](https://doc.qt.io/qt-6/qstring.html#QStringLiteral))

The value of a `QStringLiteral` is that it will create the `QString` at compilte time, instead of instantiating at every function calls. This is a macro optimisation.

The TL;DR is: if you use a static string in your `QString` (like here with `" FILE \"%1\"\n"`), you can make it a `QStringLiteral`.
You are correct that the one above could also be using `QStringLiteral` tho! Feel free to update them.

presentformyfriends replied:
Thanks for the explanation, I will update mine and the ones above.

acolombier commented on src/preferences/dialog/dlgprefrecord.cpp:
```diff
@@ -201,7 +212,12 @@ void DlgPrefRecord::slotResetToDefaults() {
     CheckBoxRecordCueFile->setChecked(kDefaultCueEnabled);
+
+    // Sets 'Enable File Annotation in CUE file' checkbox value
+    CheckBoxUseCueFileAnnotation->setChecked(false);
```
Can you capture that default value in a `kConstant` like for line 217?

presentformyfriends replied:
Sure, I did this originally and then scrapped it when I realized I could keep it as `false`. This builds fine on my end. I was curious why `kDefaultCueEnabled` was necessary in this case but I could not find it anywhere else in the code. So if you can explain why we capture the default in this way (as opposed to say `getValueString`), I would greatly appreciate it!

acolombier replied:
This helps with readability mainly -  all defaults get listed at the top of the file, and this helps to understand why this `false` is being set to `CheckBoxUseCueFileAnnotation`.

presentformyfriends replied:
That makes sense, thanks!

acolombier commented on src/preferences/dialog/dlgprefrecord.cpp:
```diff
@@ -303,6 +319,16 @@ void DlgPrefRecord::slotSliderQuality() {
 
+// Set 'Enable File Annotation in CUE file' checkbox value depending on 'Create a CUE file' checkbox value
+void DlgPrefRecord::updateCueEnabled() {
+    if (CheckBoxRecordCueFile->isChecked()) {
+        CheckBoxUseCueFileAnnotation->setEnabled(true);
+    } else {
+        CheckBoxUseCueFileAnnotation->setEnabled(false);
+        CheckBoxUseCueFileAnnotation->setChecked(false);
+    }
```
Do we have to uncheck the `File Annotation` checkbox? Disabling make sense tho, to highlight the dependency.

```suggestion
   CheckBoxUseCueFileAnnotation->setEnabled(CheckBoxRecordCueFile->isChecked());
```

acolombier commented on src/preferences/dialog/dlgprefrecord.cpp:
```diff
@@ -429,11 +455,20 @@ void DlgPrefRecord::saveEncoding() {
     }
 }
 
+void DlgPrefRecord::slotToggleCueEnabled() {
+    updateCueEnabled();
+}
```
I think these two methods are redundant. I would suggest keeping `slotToggleCueEnabled` and call it directly if you need to.

presentformyfriends replied:
True

presentformyfriends replied:
Just pushed new commit for this.

acolombier reviewed (changes requested):
This is looking good, thanks for the PR!
 Just some minor comments

presentformyfriends commented:
> you might need to do a `pre-commit run --from-ref upstream/main --to-ref HEAD` so it runs on the entire PR.

I ran this command but it still fails on clang-format and lists the same files:

`
fix utf-8 byte order marker..............................................Passed
check for case conflicts.................................................Passed
check json...........................................(no files to check)Skipped
check for merge conflicts................................................Passed
check xml............................................(no files to check)Skipped
check yaml...........................................(no files to check)Skipped
fix end of files.........................................................Passed
mixed line ending........................................................Passed
trim trailing whitespace.................................................Passed
don't commit to branch...................................................Passed
codespell................................................................Passed
eslint...............................................(no files to check)Skipped
clang-format.............................................................Failed
- hook id: clang-format
- files were modified by this hook

[INFO] First pass: Reformatting added/changed lines...
[INFO] Reformatting src/engine/sidechain/enginerecord.cpp [21-22, 40-40, 254-259]...
[INFO] Reformatting src/engine/sidechain/enginerecord.h [88-88]...
[INFO] Reformatting src/preferences/dialog/dlgprefrecord.cpp [79-81, 125-129, 155-155, 187-187, 191-192, 215-216, 218-220, 322-331, 458-461, 467-471]...
[INFO] Reformatting src/preferences/dialog/dlgprefrecord.h [35-37, 50-51]...
[INFO] Second pass: Breaking long added/changed lines...
[INFO] Reformatting src/engine/sidechain/enginerecord.cpp [40-40]...
[INFO] Reformatting src/preferences/dialog/dlgprefrecord.cpp [322-322]...

black................................................(no files to check)Skipped
flake8...............................................(no files to check)Skipped
shellcheck...........................................(no files to check)Skipped
markdownlint-cli2....................................(no files to check)Skipped
Validate GitHub Workflows............................(no files to check)Skipped
prettier.............................................(no files to check)Skipped
reformat QML files...................................(no files to check)Skipped
qsscheck.............................................(no files to check)Skipped
changelog............................................(no files to check)Skipped
qmllint..............................................(no files to check)Skipped
metainfo.............................................(no files to check)Skipped
 `

presentformyfriends committed:
Update src/preferences/dialog/dlgprefrecord.cpp

Co-authored-by: Antoine Colombier <7086688+acolombier@users.noreply.github.com>

presentformyfriends committed:
Update src/preferences/dialog/dlgprefrecord.cpp

Co-authored-by: Antoine Colombier <7086688+acolombier@users.noreply.github.com>

daschuer commented:
> I ran this command but it still fails on clang-format and lists the same files:

Did you 
`git commit -a` after checking all files?

daschuer commented:
Alternative is to apply this patch: 
https://github.com/mixxxdj/mixxx/actions/runs/9982325217/artifacts/1713542080

presentformyfriends committed:
Add files modified by clang-format hook

presentformyfriends commented:
> Did you `git commit -a` after checking all files?

No, just tried that and now I'm unable to even push or pull.

presentformyfriends committed:
Revert "Add files modified by clang-format hook"

This reverts commit 5a071634179b2039832c376ec36c8c5671b11345.

presentformyfriends committed:
Merge divergent branches

presentformyfriends commented:
Ok should be fixed now, just did a git revert on most recent commit and then pulled. Please let me know if there is anything else I need to do to complete the PR, and thanks to all for the help.

daschuer commented on src/engine/sidechain/enginerecord.cpp:
```diff
@@ -249,6 +251,12 @@ void EngineRecord::writeCueLine() {
                             .toUtf8());
 
+    if (m_bCueUsesFileAnnotation) {
+        m_cueFile.write(QString("    FILE \"%1\"\n")
```
```suggestion
        m_cueFile.write(QStringLiteral("    FILE \"%1\"\n")
```

Feel free to also fix other occurrence

presentformyfriends replied:
Just pushed new commit that should take care of this. I updated the line you suggested as well as the ones above it. Thanks!

presentformyfriends committed:
Fix line length, naming case, redundant function

Resolve line length issues to pass clang-format hook in pre-commit.
Change config key to snake_case.
Remove redundant 'updateCueEnabled' function, call slotToggleCueEnabled
directly instead.

presentformyfriends committed:
Revert "Fix line length, naming case, redundant function"

This reverts commit 77cc748df0e0abd626236433e8c3e0919a19abc5.

presentformyfriends committed:
Fix line length, naming case, redundant function

Resolve line length issues to pass clang-format hook in pre-commit.
Change config key to snake_case.
Remove redundant 'updateCueEnabled' function, call slotToggleCueEnabled
directly instead.

presentformyfriends commented:
FYI I skipped clang-format and end-of-file-fixer for the most recent commit.

acolombier reviewed (approved):
LGTM! Thanks for addressing the feature request.

It may be that the pipeline complains due to some precommit checks but otherwise, it looks good now.

presentformyfriends commented:
> LGTM! Thanks for addressing the feature request.

Nice! Thanks for the guidance and thank you to everyone in here!

> It may be that the pipeline complains due to some precommit checks but otherwise, it looks good now.

Ok good, I'm going to go read the Pro Git book again lol

acolombier committed:
Merge remote-tracking branch 'upstream/main' into fixing_some_bug

acolombier committed:
chore: address pre-commit linting

acolombier commented:
@daschuer I have addressed the pre-commit, are you happy with updating your review?

daschuer commented on src/preferences/dialog/dlgprefrecord.cpp:
```diff
@@ -201,7 +214,12 @@ void DlgPrefRecord::slotResetToDefaults() {
     CheckBoxRecordCueFile->setChecked(kDefaultCueEnabled);
+
+    // Sets 'Enable File Annotation in CUE file' checkbox value
+    CheckBoxUseCueFileAnnotation->setChecked(false);
```
```suggestion
    CheckBoxUseCueFileAnnotation->setChecked(kDefaultCueFileAnnotationEnabled);
```

daschuer reviewed (changes requested):
Thank you for adopting this. I left two final comments

Swiftb0y commented on src/engine/sidechain/enginerecord.cpp:
```diff
@@ -35,7 +36,16 @@ int EngineRecord::updateFromPreferences() {
     m_baAlbum = m_pConfig->getValueString(ConfigKey(RECORDING_PREF_KEY, "Album"));
     m_cueFileName = m_pConfig->getValueString(ConfigKey(RECORDING_PREF_KEY, "CuePath"));
-    m_bCueIsEnabled = m_pConfig->getValueString(ConfigKey(RECORDING_PREF_KEY, "CueEnabled")).toInt();
+    m_bCueIsEnabled =
+            m_pConfig
+                    ->getValueString(ConfigKey(
+                            RECORDING_PREF_KEY, "CueEnabled"))
+                    .toInt();
+    m_bCueUsesFileAnnotation =
+            m_pConfig
+                    ->getValueString(ConfigKey(
+                            RECORDING_PREF_KEY, "cue_file_annotation_enabled"))
+                    .toInt();
```
also `QStringLiteral` while ur at it ;)

acolombier committed:
Apply suggestions from code review

Co-authored-by: Daniel Schürmann <daschuer@mixxx.org>

acolombier commented:
Done @daschuer

acolombier commented:
Friendly ping @daschuer - addressed your change requests

presentformyfriends commented:
Hi did this ever get merged? I don't see the functionality in 2.5?

daschuer commented:
Not yet. This needs a final test.

Swiftb0y reviewed (approved):
Haven't done a test, but code LGTM.

Swiftb0y commented:
whoops. I left a comment but apparently it has gotten lost somehow.

I can confirm I now have a `FILE` line in the cuesheet. However unlike the top level `FILE` tag (referencing the recording), there is no indication of the filetype (like `WAVE` or `MP3`) after the path. Is this needed or optional @presentformyfriends?

daschuer reviewed (approved):
I can confirm the review issues are solved.
Edit: Oh I see @Swiftb0y has tested an his finding has been clarified.

presentformyfriends commented:
Oh ok good, thanks for clairfying that for me. Much appreciated.

daschuer commented:
@presentformyfriends I like to add you name to the contrributor list in the Mixxx about box. Which name shall I use?
